lib/tapioca/dsl/helpers/active_record_column_type_helper.rb

```diff
@@ -113,7 +115,7 @@ def column_type_for(column_name)
               getter_type
             end
 
-          if @column_type_option.persisted? && !column&.null
+          if @column_type_option.persisted? && (virtual_attribute?(column_name) || !nilable_column)
```

Comment:
The `virtual_attribute?` check here was surprising to me (I do not have much experience with these), but there was a spec enforcing it, so I made it more explicit and retained the behavior.

Reply:
Oh the default is persisted https://github.com/Shopify/tapioca/blob/f2010b52a1980b51f649ac4277f37a98f0accce5/lib/tapioca/dsl/helpers/active_record_column_type_helper.rb#L32

Reply:
Yes, sorry for the confusion. I just meant that my intuition would be that virtual attributes would be nil-able by default, but the  spec specifically enforced that they were non-nil.

Reply:
So before `!column&.null` would return `false` and we didn't enter this branch. But using `!nilable_column` we are entering this branch and you need the extra conditional correct? I'd like to understand why this difference is happening and make sure it's not a bug in the new implementation. Is there an implicit presence validator in the test model?

Reply:
Hi @KaanOzkan . Yes, this is super subtle and confused me as well. 

Put succinctly, the confusing behavior is that `!column&.null` is `true` if `column` is `nil`, which is the case if the attribute is virtual. This means that we were always falling into the branch for virtual attributes. I don't know whether this is desirable, except that there was a spec relying on this behavior. 

In my code, `has_non_null_database_constraint?` is only `true` if a column exists because we return `false` early if the `column` is `nil`, thus separating out the virtual and non-virtual cases (specifically whether a `column` is found or not), so I need to add an additional explicit check (`virtual_attribute?`) that checks for `nil`-ness of the `column` to retain the existing behavior. I believe this is ultimately clearer and easier to follow.

I hope that helps clarify things.

Reply:
Thank you for the explanation. I'm not sure if the test is enforcing this or it just happened to follow this code path. cc @bdewater

I'm okay with keeping the current behaviour but I'd prefer if it was cleaner. Could we remove `return false if column.nil?` instead in `has_non_null_database_constraint`?

Reply:
> I'm okay with keeping the current behaviour but I'd prefer if it was cleaner. Could we remove return false if column.nil? instead in has_non_null_database_constraint?

I'm happy to make this change if you so prefer, but to me, it's cleaner to make the two conditions explicit and separate. I believe the reason it was confusing in the current code is that they were subtly intertwined. Additionally, it seems to me that a method named `has_non_null_database_constraint?` should be `false` if there is no database column to be constrained. 

Again, happy to make the change to align with the rest with the project and its maintainers standards and preferences, so let me know if my logic falls short for you.

Reply:
> It felt too specific of a concept for type helper to know but I think you're right. I'll wait on @egiurleo's review before shipping. 

Sounds good, I appreciate the time and thoughts here!

> Also there's second commit doesn't have a message, I think it can be squashed in to the first.

Ah yes, we use squash and commit here at AL, and I forget that not everyone does that. Let me update the branch.